src/pages/index.js

```diff
@@ -38,76 +48,143 @@ const section = new Section(
-	selectors.modal.profileEdit.modal,
-	handleProfileFormSubmit
+
+const addNewCardFormValidator = new FormValidator(
+	configValidation,
+	addNewCardForm
 );
 
-function handleProfileFormSubmit(inputValues) {
-	const { name, about } = inputValues;
-	userInfo.setUserInfo({ name, job: about });
-	profileEditPopup.close();
-}
+const avatarEditFormValidator = new FormValidator(
+	configValidation,
+	avatarEditForm
+);
+
+const cardPreviewPopup = new PopupWithImage(selectors.modal.image.modal);
+
+// --------------------------------------------------------------//
+// Code flow--------- -------------------------------------------//
+// --------------------------------------------------------------//
+
+api.getUserInfoAndCards().then(([userData, initialCards]) => {
+	userInfo.setUserInfo(userData);
+	section.renderItems(initialCards);
+});
 
 profileEditPopup.setEventListeners();
+avatarEditPopup.setEventListeners();
+addNewCardPopup.setEventListeners();
+cardPreviewPopup.setEventListeners();
+
 profileEditButton.addEventListener(eventType.CLICK, openEditProfileModal);
+addNewCardButton.addEventListener(eventType.CLICK, openCardAddModal);
+avatarElem.addEventListener(eventType.CLICK, openProfilePicModal);
+
 profileEditFormValidator.enableValidation();
+addNewCardFormValidator.enableValidation();
+avatarEditFormValidator.enableValidation();
+
+// --------------------------------------------------------------//
+// Reusable functions -------------------------------------------//
+// --------------------------------------------------------------//
+function openCardAddModal() {
+	addNewCardFormValidator.resetValidation();
+	addNewCardPopup.open();
+}
+
+function openProfilePicModal() {
+	const { avatar } = userInfo.getUserInfo();
+	avatarInput.value = avatar;
+	avatarEditFormValidator.resetValidation();
+	avatarEditPopup.open();
+}
 
 function openEditProfileModal() {
-	const { name, job } = userInfo.getUserInfo();
+	const { name, about } = userInfo.getUserInfo();
 	profileTitleInput.value = name;
-	profileDescriptionInput.value = job;
+	profileAboutInput.value = about;
 	profileEditFormValidator.resetValidation();
 	profileEditPopup.open();
 }
 
-// POPUP: CARD ADD FORM
-const addNewCardFormValidator = new FormValidator(
-	configValidation,
-	addNewCardForm
-);
+function handleCardPreviewClick(caption, imageUrl) {
+	cardPreviewPopup.open(caption, imageUrl);
+}
 
-const addNewCardPopup = new PopupWithForm(
-	selectors.modal.cardAdd.modal,
-	handleNewCardSubmit
-);
+function handleProfileFormSubmit(inputValues) {
+	const { name, about } = inputValues;
+	const submitButtonElem = profileEditForm.querySelector(
+		selectors.modal.button
+	);
+	const oldButtonLabel = submitButtonElem.textContent;
+	submitButtonElem.textContent = 'Saving...';
+	api.editUserInfo({ name, about })
+		.then((userData) => {
+			userInfo.setUserInfo(userData);
+			profileEditPopup.close();
+		})
+		.finally(() => {
+			submitButtonElem.textContent = oldButtonLabel;
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```